src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,197 @@
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = tail = new Node<>(null, value, null);
+        } else {
+            tail = tail.next = new Node<>(tail, value, null);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        addByIndexRoutine(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkSetIndexInBound(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        return setValueRoutine(value, index);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlinkNodeByIndex(index);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return unlinkNodeByObject(object);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private void addByIndexRoutine(T value, int index) {
+        checkIndexInBound(index);
+        if (head == null) {
+            head = tail = new Node<>(null, value, null);
+        } else if (index == 0) {
+            head = new Node<>(null, value, head);
+        } else if (index == size) {
+            tail = tail.next = new Node<>(tail, value, null);
+        } else {
+            Node<T> currentNode = getNode(index - 1);
+            Node<T> node = new Node<>(currentNode, value, currentNode.next);
+            currentNode.next.prev = node;
+            currentNode.next = node;
+
+        }
+        size++;
+    }
+
+    private T setValueRoutine(T value, int index) {
+        checkSetIndexInBound(index);
+        Node<T> nodeToSet;
+        if (index == 0) {
+            nodeToSet = head;
+        } else {
+            nodeToSet = getNode(index - 1).next;
+        }
+        T returnValue = nodeToSet.value;
+        nodeToSet.value = value;
+        return returnValue;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current;
+        if (index < size / 2 + 1) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i >= index + 1; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (isCurrentNodeValueEquals(value, current)) {
+                return current;
+            }
+            current = current.next;
+        }
+        return current;
```

Comment:
you are returning head if you did not find the node by value, we should return null instead, because it is misinformative
```suggestion
        return null;
```